File: iosApp/LucidDreaming/ViewModels/MonitorViewModel.swift
import Foundation
import Combine
import Shared

class MonitorViewModel: ObservableObject {
    @Published var gameInfo: GameInfo? = nil
    @Published var isLoading: Bool = false
    @Published var error: String? = nil
    @Published var screenshot: Data? = nil
    
    private let repository = GameInfoRepository()
    private let settingsPreferences = iOSSsettingsPreferences()
    private var cancellables = Set<AnyCancellable>()
    private var autoRefreshEnabled = false
    private var refreshTask: Task<Void, Never>? = nil
    
    func loadGameInfo() {
        isLoading = true
        error = nil
        
        Task {
            do {
                let result = try await repository.getGameInfo()
                switch result {
                case .success(let gameInfo):
                    self.gameInfo = gameInfo
                case .failure(let error):
                    self.error = error.localizedDescription
                }
                
                isLoading = false
            } catch {
                self.error = error.localizedDescription
                isLoading = false
            }
        }
    }
    
    func loadScreenshot() {
        isLoading = true
        
        Task {
            do {
                let result = try await repository.getScreenshot()
                switch result {
                case .success(let screenshotData):
                    screenshot = screenshotData
                case .failure(let error):
                    self.error = error.localizedDescription
                }
                
                isLoading = false
            } catch {
                self.error = error.localizedDescription
                isLoading = false
            }
        }
    }
    
    func clearError() {
        error = nil
    }
    
    func startAutoRefresh() {
        guard !autoRefreshEnabled else { return }
        autoRefreshEnabled = true
        
        refreshTask = Task {
            while autoRefreshEnabled {
                let settings = await settingsPreferences.settings.first(where: { _ in true })
                let interval = settings?.refreshInterval ?? 10
                
                try? await Task.sleep(nanoseconds: UInt64(interval) * 1_000_000_000)
                
                if autoRefreshEnabled {
                    loadGameInfo()
                }
            }
        }
    }
    
    func stopAutoRefresh() {
        autoRefreshEnabled = false
        refreshTask?.cancel()
        refreshTask = nil
    }
}

File: androidApp/src/main/java/com/luciddreaming/android/viewmodel/MonitorViewModel.kt
package com.luciddreaming.android.viewmodel

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.luciddreaming.shared.data.model.GameInfo
import com.luciddreaming.shared.data.preferences.AndroidSettingsPreferences
import com.luciddreaming.shared.data.preferences.AppSettings
import com.luciddreaming.shared.data.repository.GameInfoRepository
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import kotlinx.coroutines.delay

class MonitorViewModel(application: Application) : AndroidViewModel(application) {
    private val repository = GameInfoRepository()
    private val settingsPreferences = AndroidSettingsPreferences(application)

    private val _gameInfo = MutableStateFlow<GameInfo?>(null)
    val gameInfo: StateFlow<GameInfo?> = _gameInfo.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    private val _screenshot = MutableStateFlow<ByteArray?>(null)
    val screenshot: StateFlow<ByteArray?> = _screenshot.asStateFlow()

    private var autoRefreshEnabled = false

    fun loadGameInfo() {
        viewModelScope.launch {
            _isLoading.value = true
            _error.value = null

            val result = repository.getGameInfo()
            if (result.isSuccess) {
                _gameInfo.value = result.getOrNull()
            } else {
                _error.value = result.exceptionOrNull()?.message ?: "Failed to load game info"
            }

            _isLoading.value = false
        }
    }

    fun loadScreenshot() {
        viewModelScope.launch {
            _isLoading.value = true

            val result = repository.getScreenshot()
            if (result.isSuccess) {
                _screenshot.value = result.getOrNull()
            }

            _isLoading.value = false
        }
    }

    fun clearError() {
        _error.value = null
    }

    fun startAutoRefresh() {
        if (autoRefreshEnabled) return
        autoRefreshEnabled = true

        viewModelScope.launch {
            while (autoRefreshEnabled) {
                val settings = settingsPreferences.settings.first()
                val interval = settings.refreshInterval

                delay(interval * 1000L)

                if (autoRefreshEnabled) {
                    loadGameInfo()
                }
            }
        }
    }

    fun stopAutoRefresh() {
        autoRefreshEnabled = false
    }
}
